Use `QuantityValue` with fractions for exact unit conversions

lipchev opened the pull request:
## Summary

Replaces floating-point quantity storage and generated conversion switches with exact, fraction-backed values and conversion expressions.

- Reworks `QuantityValue` as a rational number with standard numeric interop, arithmetic, parsing and formatting.
- Uses `QuantityValue` throughout generated quantities, operators, unit relations and conversion APIs.
- Generates exact unit conversion expressions and redesigns `UnitConverter` with configurable caching, custom quantities and quantity-to-quantity conversions.
- Updates Json.NET value formats and adds the `UnitsNet.Serialization.SystemTextJson` package with quantity, unit and value converters.
- Adds migration documentation, configuration samples, tests and benchmarks.

## Compatibility

This is a v6 breaking change: quantity values and conversion results now use `QuantityValue`; `UnitConverter` configuration APIs changed; and some serialized value representations change. Numeric inputs remain convenient through implicit conversions, and `QuantityValue` converts implicitly to `double` where a floating-point value is required.

See the [v5 to v6 upgrade guide](https://github.com/angularsen/UnitsNet/blob/fractional-quantity-value/Docs/upgrading-from-5.x-to-6.x.md) and [serialization guide](https://github.com/angularsen/UnitsNet/blob/fractional-quantity-value/Docs/serialization.md).

## Validation

- Full solution build
- 52,818 main-library tests passed on .NET 10 (16 skipped)
- Json.NET tests passed on .NET Framework 4.8 and .NET 10
- System.Text.Json tests passed on .NET Framework 4.8 and .NET 8, 9 and 10

lipchev committed:
- QuantityValue implemented as a fractional number
- IQuantity interfaces optimized (some methods refactored as extensions)
- QuantityInfo/UnitInfo hierachy re-implemented (same properties, different constructors)
- QuantityInfoLookup is now public
- UntAbbreviationsCache, UnitParser, QuantityParser optimized
- UnitConverter: re-implemented (multiple versions)
- removed the IConvertible interface
- updated the JsonNet converters
- introducing the SystemTextJson project
- added a new UnitsNetConfiguration to the Samples project showcasing the new configuration options
- many more tests and benchmarks (perhaps too many)

lipchev commented:
@angularsen Clearly, I don't expect this to get merged in the _Gitty up!_ fashion, but at least we have the whole picture, with sources that I can reference.

If you want, send me an e-mail, we could do a quick walk-through / discussion.

lipchev committed:
adding the git-ignored files

lipchev committed:
changing the CultureInfo? parameters back to IFormatProvder?

lipchev committed:
- `UnitAbbreviationsCahe` and `UnitParser`: replaced the IReadOnlyCollection constructors with IEnumerable
- `UnitAbbreviationsCacheInitializationBenchmarks`: replaced some obsolete usages

angularsen commented:
100k lines removed, 100k lines added 🙈 

![image](https://github.com/user-attachments/assets/b600c4ac-ce3c-4417-9171-a9b9782d88f1)

lipchev commented:
> 100k lines removed, 100k lines added 🙈

I tried to create this PR twice before (many months ago), while the changes to the unit definitions were still not merged- and the web interface was giving me an error when trying to browse the files changed.. Something like "Too many files to display" 😄

angularsen commented:
Ok, I'm not going to get through a review of this many files anytime soon.
Maybe we should have a screen sharing session and go through it together. I may have some time this weekend, what about you? What timezone are you in?

On the surface though, it seems like this could be split up into chunks. I know it's tedious and extra work, but it will be way faster to review. Do you see any chunks of changes to easily split off into separate PRs?

lipchev commented:
> Ok, I'm not going to get through a review of this many files anytime soon. Maybe we should have a screen sharing session and go through it together. I may have some time this weekend, what about you? What timezone are you in?

Sofia (GMT+3), but time zones are not relevant to my sleep schedule - so basically any time you want.
>
> On the surface though, it seems like this could be split up into chunks. I know it's tedious and extra work, but it will be way faster to review. Do you see any chunks of changes to easily split off into separate PRs?

Yes, I do have some ideas:
1. `UnitAbbreviationsCache` and the `UnitParser` should be more or less free of changes once #1548 is merged
2. I plan to remove the `IConvertible` interface tonight (lots of _red_ points there)
3. The `QuantityParser` has just a few minor changes which I was going to try to push as well (other than that it's mostly just `double` changing to `QuantityValue`)
4. `QuantityFormatter` - there was an issue that I created earlier that should (mostly) solve the differences
5. Refactoring the `QuantityInfo` can theoretically be done without the `ConversionExpression`s (which would open the way for the changes to the `IQuantity` interface and some of the extension methods).
6. The `JsonQuantityConverter` stuff from `SystemText` could theoretically come with it's `double` versions first (but we do need to have a discussion about it)

Hopefully by the time we get to 5) you'd be up to speed (and fed up with PRs) and we can turn back to reviewing / working on _the rest of it_ as a whole 😄

angularsen commented:
Ok, sounds good. Just send PRs my way and I'll try to get to them. I have a little bit of extra time this weekend.

lipchev committed:
including the changes from https://github.com/angularsen/UnitsNet/pull/1548/commits/f9a2766b959334432c531c44afca207f37c155cc

lipchev committed:
Completed the QuantityFormatter test coverage (and some cosmetic changes)

lipchev committed:
UnitKey: making the constructor internal and renaming the properties

lipchev committed:
removing the QuantityFormatter overloads and re-organizing the extensions into their own folder

lipchev committed:
- updating the implicit usings and cleanup some of the using statements
- UnitAbbreviationsCache: removing the obsolete overload
- adding a few more tests

lipchev committed:
adding a few more tests

lipchev committed:
removing some unnecessary using statements and replacing the target framework (net8.0) with 9.0 on the QuantityInfoBenchmarks

lipchev committed:
renaming the default quantity provider

lipchev committed:
removing Quantity.FromQuantityInfo

lipchev committed:
avoiding the ToArray call when initializing the quantity info using an arrray (as IEnumerable)

lipchev committed:
Merge remote-tracking branch 'upstream\master' into fractional-quantity-value

angularsen commented:
Great! I may have some time this weekend to look at it.

lipchev committed:
Merge remote-tracking branch 'upstream\master' into fractional-quantity-value

lipchev committed:
removing the remnents of the IQuantityInfo<TUnit> interface and the unused constructors of the UnitInfo

lipchev committed:
removed the unit-not-found comment

lipchev committed:
merging the removal of the EnumUtils (i'm keeping it for now)

lipchev committed:
merging the AoT compatibility: https://github.com/angularsen/UnitsNet/commit/6abb5f4f086529d44897e7a2a102519140a8843d

lipchev committed:
adding the missing TryGetUnitFromAbbreviation overload and the tests

lipchev committed:
merging from upstream: ✨Add more Gray units (#1568)

lipchev committed:
removed  the EnumUtils and replacing the usages with the EnumHelper

lipchev committed:
bumping the pre-release version to 6.0.0-pre017 (staying ahead of the officially published nugget)

lipchev committed:
merging from upstream (keeping ahead of the official nugget version: 6.0.0-pre016)

lipchev committed:
cleanup some namespaces and other cosmetic changes

lipchev committed:
merging from upstream and bumping the nugget versions (keeping ahead of the official nugget version: 6.0.0-pre017)

lipchev committed:
splitting the IQuantity interface definitions into separate files

lipchev committed:
- renamed IQuantityInstance to IQuantityOfType
- replaced the type of the tolerance parameter of the Equals extension with the concrete quantity type

lipchev committed:
making the GenericMathExtensions class obsolete and replacing the relevant usages from the benchmarks

lipchev committed:
remove the GenerateArithmetic from the Temperature.json

lipchev committed:
- re-formatting the QuantityInfo and making the IQuantityInfo internal in net8+
- cleaning up some of the old benchmarks

lipchev committed:
cleanup and reformat the UnitInfo/IUnitDefinition

lipchev committed:
merging from upstream: Add prefix milli to torr unit (#1588)

lipchev committed:
cosmetic changes

lipchev committed:
refactored test: From_ValueAndUnit_ReturnsQuantityWithSameValueAndUnit

lipchev committed:
renamed the UnitsNetSetup.QuantityInfoLookup

lipchev committed:
Merge remote-tracking branch 'upstream\master' into fractional-quantity-value

lipchev committed:
synchronizing the changes in UnitKey

lipchev committed:
merged from upstream (ignoring the ToUnitUntyped bits)

lipchev committed:
- Constraint the Abs extension method to ILinearQuantity<TQuantity>
- fix the comments for the UnitMath.Clamp

lipchev committed:
merged from upstream

lipchev committed:
updating the nugget packages

lipchev committed:
refactored the [Obsolete] IQuantity methods as extensions

lipchev committed:
exposing the operator IDivisionOperators<TSelf, TSelf, QuantityValue>

lipchev committed:
bump the nugget version to 6.0.0-pre019

lipchev committed:
making the TQuantity Create(double, UnitKey) method public (IQuantityOfType / IQuantityInstanceInfo)

lipchev committed:
removing the Dimensions instance property

lipchev committed:
merged from upstream

lipchev committed:
removed the Dimensions from the HowMuch

lipchev committed:
completing the implementation of the HowMuch in the Configuration sample

lipchev committed:
merged from upstream

angularsen commented:
@lipchev I really like the concept of the decimal fractions, it solves equality issues and precision, and if I recall correctly it remains fairly backwards compatible. 

I want to say let's go, but if you could please help me out and give a short summary for a couple of things that would help a lot as I don't recall and I'm pressed for time.

- How does it affect performance? CPU time on creating a quantity, converting units and ToString are perhaps the big ones. Allocations are also relevant. I don't need full benchmarks, just a top level understanding of are we talking 2x, 10x, 100x orders of magnitude.
- Any significant breaking changes to be aware of?

lipchev commented:
>     * How does it affect performance? CPU time on creating a quantity, converting units and ToString are perhaps the big ones. Allocations are also relevant. I don't need full benchmarks, just a top level understanding of are we talking 2x, 10x, 100x orders of magnitude.

I spent a full Sunday back in February creating a comparison chart, but before I could finish up setting up all the sheets, I accidentally executed the build script which erased all of my benchmark results. 😠 I'll attach the file as it is, but the gist is that compared to `v5` there should be a significant performance _improvement_, especially when it comes to the strings. 
[UnitsNet-Benchmarks.xlsx](https://github.com/user-attachments/files/22478160/UnitsNet-Benchmarks.xlsx)

The conversions are about the same, or faster (with units other than the `BaseUnit`). There are no allocations, as long as the `Value` is represented by a fraction, who's terms are smaller than `int.MaxValue`. 

For example the value `42.24` is represented as `4224 / 100`, which when multiplied by `1000` (some `Kilo` conversion) would become `4224000 / 100` which is still small enough. Note that I intentionally avoid the (possible) reduction, since keeping the original denominator generally makes the additions / substance more efficient: for example adding `1.23` to would result in `4224123 / 100` (simply adding the integers in the numerators). So, while dividing `4224000 / 100` by 1000 would revert back to `4224 / 100`, doing a bunch of multiplications (without manually calling `QuantityValue.Reduce`) would sooner or later hit the magical threshold (`int.MaxValue`) which would trigger the allocation of the `uint[]? _bits`. From there on, the performance scales linearly with the size of the bits increasing by 1 for every power of `int.MaxValue` (my PC was still able to do square roots, even after something like 500K digits of precision). Anyway, I've got a ton of benchmarks of the implementation over at the `Fractions` repo (with loads of pretty charts) - when we factor in the occasional allocations, the raw arithmetic-performance was within the same order of magnitude as the `decimal` (it was [2x-6x slower](https://github.com/Fractions-DotNet/Fractions/pull/72#issuecomment-2171429662) back on Jul 1, 2024, and I expect the `QuantityValue` to be slightly faster). 

The initialization is slower but no more than 10x, judging from [these benchmarks](https://github.com/angularsen/UnitsNet/pull/1543) but that can be reduced by loading a select subset of quantities (I imagine loading just 10 `QuantityInfo`s would still at least 2x faster). 

PS The `Inverse` method is also something like 2x slower (but that's due to the extra logic)...

PS2 Aside for the static benchmarks, I've also observed a visible improvement (in the performance logs) after migrating my code-base, but that's probably due to the improvements to the string handling stuff around the `UnitParser`, `UnitAbbreviations` etc. (from `v5` to `v6`). 

>     * Any significant breaking changes to be aware of?

Compared to the `v5` the only significant change is that
- `double value = Mass.Zero.Value` is not allowed
- `decimal value = Power.Zero.Value` is not allowed
both of these return a `QuantityValue` which does not support the implicit conversion to `double` or `decimal`:
- `double value = (double)Mass.Zero.Value` is allowed
- `decimal value = Power.Zero.Value.ToDecimal()` is also allowed

While inconvenient, this is the _correct way_ to do it (implicit conversions from, and explicit conversion to any primitive number).

By the way, given that in `v5` `IQuantity.Value` is still defined as `QuantityValue` (without an explicit cast as far as I remember), this shouldn't be _that much of surprise_ 🤷

angularsen committed:
Merge remote-tracking branch 'origin/master' into fractional-quantity-value

angularsen committed:
Samples/Nuget.config: Use artifcats folder to access all locally built nugets

angularsen committed:
samples/UnitConverter.Wpf: Use double example values

angularsen committed:
samples: Add 'Official' solution configuration, remove all but AnyCPU

On Rider, this was needed to actually run the samples with 'Official' configuration. Only Debug/Release were shown.

angularsen committed:
Revert "samples/UnitConverter.Wpf: Use double example values"

This reverts commit 896f7653a16f38f9b6c164dbc5198bdf21c3f8a4.

angularsen committed:
samples: ⬆️UnitsNet 5.75.0

angularsen reviewed (commented):
Posting review so far, I got down through SystemTextJson.

angularsen committed:
♻️Rename to 'expected' for some test cases

angularsen committed:
♻️Rename to WithConversionFactorFrom/ToBase

angularsen committed:
♻️Add parameter names for ConversionExpression ctor arguments

angularsen committed:
⚡Use GeneratedRegex in ExpressionEvaluator

angularsen committed:
🐛Fix fraction parsing on non-English system, add FractionHelper.TryParseInvariant

Specify invariant culture for Fraction and `int` parsing.

angularsen committed:
♻️Specify invariant culture for all int parsing

This is typically not a problem with integers, but let's be explicit.
Integer parsing is done for quantity string formatting to control number formats, such as `a2` for the 2nd unit abbreviation.

angularsen committed:
♻️Fix whitespace in generated quantities

angularsen committed:
♻️Fix inconsistent ordering of UnitRelations.json, use OrdinalIgnoreCase

angularsen reviewed (commented):
Another batch of comments.

I got down to: `ConversionExpressionTests.cs`

https://github.com/angularsen/UnitsNet/pull/1544/files?show-viewed-files=false#diff-98dfe7c666a4282deaf6c6e7e315718f09abf08b96d63319d413a3f362102b36

angularsen committed:
♻️Use named arguments for significantDigits

lipchev committed:
- adding net10.0 to the list of target frameworks
- updating all packages to their latest versions
- UnitsNet version bumped to 6.0.0-pre100

lipchev committed:
fix(debug): move GetConfiguration call to property getter to avoid ReSharper debugger failure
Move the configuration lookup out of the QuantityDebugProxy constructor into the Configuration property getter in UnitsNet/CustomCode/Debug/QuantityDebugProxy.cs so the ReSharper inline evaluator no longer shows "Cannot evaluate expression" when expanding the proxy in the debugger. Also replace debug-formatting static fields with properties (DefaultFormatSpecifier, DefaultFormatProvider) and tidy namespaces/XML docs. Mitigates ReSharper issues RSRP-499956 and RSRP-502262.

lipchev committed:
Adding some comments in the QuantityGenerator regarding the UnitInfo constructor logic and improving the xml comments regarding the equality contract

lipchev commented on UnitsNet.Serialization.JsonNet/UnitsNet.Serialization.JsonNet.csproj:
```diff
@@ -1,8 +1,8 @@
   <PropertyGroup>
     <PackageId>UnitsNet.Serialization.JsonNet</PackageId>
-    <Version>6.0.0-pre007</Version>
+    <Version>6.0.0-pre019</Version>
```
fixed in [510afe83edc0a572ee32d1b1c7968eac06537f8c ](https://github.com/lipchev/UnitsNet/commit/5e88d2024cb2fa40d65311ab19f8288f2bbff31f) - UnitsNet version bumped to `6.0.0-pre100` (and added `.net10.0` to targets).

lipchev commented on CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:
```diff
@@ -895,97 +972,82 @@ private void GenerateEqualityAndComparison()
 
-        /// <summary>Indicates strict inequality of two <see cref=""{_quantity.Name}""/> quantities, where both <see cref=""Value"" /> and <see cref=""Unit"" /> are exactly equal.</summary>
-        [Obsolete(""For null checks, use `x is null` syntax to not invoke overloads. For equality checks, use Equals({_quantity.Name} other, {_quantity.Name} tolerance) instead, to check equality across units and to specify the max tolerance for rounding errors due to floating-point arithmetic when converting between units."")]
+        /// <summary>Indicates strict inequality of two <see cref=""{_quantity.Name}""/> quantities.</summary>
```
fixed in https://github.com/lipchev/UnitsNet/commit/5e88d2024cb2fa40d65311ab19f8288f2bbff31f

lipchev committed:
removing some commented out bits (as per Claud's suggestions)

angularsen commented on CodeGen/CodeGen.csproj:
```diff
@@ -1,15 +1,16 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>net9.0</TargetFramework>
+    <TargetFramework>net10.0</TargetFramework>
```
Let's do net10.0 as a separate PR, but we'll definitely bump

angularsen replied:
Handled separately on master as requested: CodeGen was retargeted to .NET 10 in commit [`e3b6adfd8`](https://github.com/angularsen/UnitsNet/commit/e3b6adfd8cd77185787a1197ca5ea734ebed3515), outside the fractional-value feature changes.

angularsen commented:
Just to give a life sign, I'm still very busy with work and I expect this to continue well into the next year. However, I'm motivated to get this PR merged in the upcoming Christmas season, bump v6 to beta and start stabilizing and collecting feedback on the rather big changes for a little while before we fully release it.

I see some of my comments haven't been addressed yet, so I assume you are still working on this? 
Let me know if you are waiting for me on anything. I'll try to help nudge this PR along in the next weeks, I believe it's close to ready to merge.

I realize I'm becoming a bottleneck in this project, we should discuss options to improve that. Maybe onboard more contributors, and I think you should have more autonomy to make decisions without having to wait on me. I'm open to ideas here.

lipchev commented on CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:
```diff
@@ -71,13 +72,13 @@ namespace UnitsNet
         /// </summary>
-        [DataMember(Name = ""Value"", Order = 1)]
-        private readonly double _value;
+        [DataMember(Name = ""Value"", Order = 1, EmitDefaultValue = false)]
```
@angularsen your call here

angularsen replied:
Addressed: `EmitDefaultValue = false` was removed from both generated data members, so `Value` and `Unit` are always included, including for `Length.Zero`.

lipchev commented on CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:
```diff
@@ -147,7 +147,7 @@ private void GenerateInterfaceExtensions()
                     }
 
-                    Writer.WL($"<{relation.LeftQuantity.Name}, {relation.RightQuantity.Name}, {relation.ResultQuantity.Name}>,");
+                    Writer.WL($"<{relation.LeftQuantity.Name}, {relation.RightQuantity.Name}, {relation.ResultQuantity.Name.Replace("double", "QuantityValue")}>,");
```
@angularsen This is an easy fix (renaming type in `UnitRelations.json`), but a potential breaking change for other source generators. If you're ok with it.. (alternatively I could potentially replace the current workaround with an extension / property).

lipchev replied:
fixed in https://github.com/angularsen/UnitsNet/commit/3c0015807cb7116349ef9ae2b655220f2efc8bf5

lipchev committed:
merged from upstream

lipchev committed:
- bumped the version to 6.0.0-pre101
- added AoT compatibility to the NumberExtensions.CS14 project

lipchev committed:
- Added new custom quantity types to the testing project: `ClassOfAffineQuantity`, `ClassOfLinearQuantity`, and `ClassOfLogarithmicQuantity`, implementing the respective interface defition (as a class).
- Completed the coverage of the `LinearQuantityExtensions`, `LogarithmicQuantityExtensions`, `AffineQuantityExtensions` (calling the concrete `Equals` overload with a null reference type)
- Completed the coverage of the `UnitMath` extensions (`Min`, `Max`)
- Cleaned up the legacy code in `IQuantityTests.cs`.

lipchev committed:
CodeGen: simplified the FractionExtensions and added some comments

lipchev committed:
- UnitRelations.json: replaced the result of the SpecificEnergy.JoulePerKilogram * BrakeSpecificFuelConsumption.KilogramPerJoule expression (`double`  - > `QuantityValue`)
- CodeGen: replaced all mentions of the `double` type

lipchev committed:
JsonNet: QuantityValueExtensions split into two files

lipchev commented on UnitsNet.Serialization.JsonNet/QuantityValueExtensions.cs:
```diff
@@ -0,0 +1,282 @@
+
+namespace UnitsNet.Serialization.JsonNet;
+
+internal static class QuantityValueExtensions
```
fixed in https://github.com/angularsen/UnitsNet/commit/8d0264ee79b7276ecb430f1f3514cb1e556e3687
PS I actually wanted to use an extension block here (one per file), but that caused the `ILLink.RoslynAnalyzer` to fail with the same exception as mentioned here https://github.com/angularsen/UnitsNet/pull/1640#discussion_r2596597579

lipchev commented on CodeGen/Helpers/ExpressionAnalyzer/Functions/Math/FractionExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+namespace CodeGen.Helpers.ExpressionAnalyzer.Functions.Math;
+
+/// <summary>
+///     We should try to push these extensions to the original library (working on the PRs)
```
fixed in https://github.com/angularsen/UnitsNet/commit/3467189d2e926c77577a23cf9b0e993a9d792756

lipchev commented on UnitsNet.Tests/CustomCode/IQuantityTests.cs:
```diff
@@ -1,84 +1,112 @@
+        //
+        //     Assert.Equal(0.0508, inSI.Value);
+        //     Assert.Equal(LengthUnit.Meter, inSI.Unit);
+        // }
```
fixed in  https://github.com/angularsen/UnitsNet/commit/576031b7d7276c41a6448cca356a135e566e92f3
Please note the 3 new test-types I've added (`ClassOfAffineQuantity`, `ClassOfLinearQuantity`, and `ClassOfLogarithmicQuantity`)- I'm just using them for one of the `Equals_TQuantity_WithNullOther_ReturnsFalse` tests,  however I think there were also other tests which require a reference type in order to tests (and I wanted to make sure that there isn't an ambiguity with the overloads).

lipchev committed:
removed some commented-out benchmark

lipchev committed:
Updated the NuGet packages

lipchev committed:
re-generated the UnitRelations.json (and fixed the "value" parameter name)

angularsen commented:
Just a heads up that I have taken small steps to resume review of this PR, I have a local review with the help of AI that I'll be iterating on a little before I post it here.

angularsen commented:
Alright, it's a bit rough and straight from the AI still but here are some new inputs on hopefully the latest changes. I haven't double checked if maybe some of these are already addressed. 

The big one is the implicit conversion, I really do think we need to reduce the blast radius of this breaking change and not require everyone to explicitly cast to double, or to deal with QuantityValue in their codebase.


# PR #1544 Review: QuantityValue as Fractional Number

**Review date**: 2026-03-09 | **Branch**: `fractional-quantity-value`

---

## Changes Required Before Merge

### P0 — Must do

1. **Add implicit conversion `QuantityValue` → `double`**
   Most consumers just want a `double`. Without this, ~60-70% of consumer code breaks on upgrade. With implicit conversion, that drops to ~20-30%. Those who want exact precision can still use `QuantityValue` directly.

2. **Revert `EmitDefaultValue = false` on DataMember**
   Remove the `EmitDefaultValue = false` parameter from `[DataMember]` on `_value` and `_unit` fields in generated quantities. `Length.Zero` should always serialize with both Value and Unit present.

### P1 — Should do

3. **Document DataContract serialization as a breaking change**
   State in v6 release notes that DataContract serialization format changed (XML and JSON). XML surrogate exists and works. DataContractJsonSerializer surrogate is blocked by a .NET runtime bug — recommend migrating to JsonNet or System.Text.Json packages.

---

## Deferred (Post-Merge)

### Before stable release

4. **Create migration guide** for breaking changes in this PR. Will need re-evaluation for full v6 migration guide later.
5. **Add test coverage for `InterfaceQuantityWithUnitTypeConverter`** — currently appears unused/untested.

### Low priority

6. **QuantityGenerator.cs** — Add example code comments in generated code sections for readability.
7. **QuantityValueFormatOptions.cs** — Align serialization/deserialization enum names (`DecimalPrecision` vs `ExactNumber` should use consistent naming).
8. **QuantityInfoBuilderExtensions** — `Configure` extension on `UnitDefinition[]` is awkward API; consider wrapper type or static method.

### Won't do

- Roslyn analyzer for migration assistance.

---

## Reference: Breaking Changes Summary

| Change | Before (v5) | After (v6) |
|--------|-------------|------------|
| `IQuantity.Value` | `double` | `QuantityValue` |
| Unit properties (e.g. `.Meters`) | `double` | `QuantityValue` |
| `Length / Length` | `double` | `QuantityValue` |
| `As()`, `ToUnit()` | Interface methods | Extension methods (some `[Obsolete]`) |
| `UnitConverter()` constructor | Public, parameterless | Removed; use `UnitConverter.Create(...)` |
| `SetConversionFunction` / `GetConversionFunction` | Available | Removed |
| `UnitsNetSetup` constructor | Public | Private; use builder pattern |
| DataContract serialization | `double` field | `QuantityValue` struct (numerator/denominator) |
| `AbbreviatedUnitsConverter` (JsonNet) | `IReadOnlyDictionary` constructor | `UnitParser` + `QuantityValueFormatOptions` |
| Default JSON precision | ~17 significant digits | Up to 29 significant digits |
| `MissingMemberHandling.Error` | Silently skipped unknowns | Now correctly throws (bug fix) |
| Null `IQuantity` deserialization | Returns `.Zero` | Returns `null` |
| Conversion factors | Floating-point approximations | Exact rational fractions |

lipchev commented:
@angularsen I'm sorry, but P0 is a no-go. I have been very explicit about it since the very beginning (even before this PR). 
You could try changing it (locally), just to get an idea of how many things would break (ambiguous assignments).

angularsen commented:
I see, I'll try and change it locally to get a better feel for why adding implicit cast is challenging. 
The way I understand this breaking change and how it will affect consumers' code, it currently feels too invasive of a change, but I'll need to map out those use cases to know for sure.

angularsen commented:
Here is a follow up, trying to map out the two alternatives to avoid or reduce the breaking change of consumer code, while retaining the value of this PR.

I had Claude implement each one and evaluate the impact on internal code and tests, as well as consunmer code.

As far as I can tell, option A with bi-directional implicit cast of QV and double is very much feasible?


# PR 1544 Review: QuantityValue as Fractional Number

The architectural improvements (ConversionExpression, centralized UnitConverter, builder pattern, exact rational coefficients) are solid and we want to get this merged.

The main concern is the breaking change impact of replacing `double` with `QuantityValue` across the public API. We evaluated two approaches to mitigate this and implemented both to get empirical data.

---

## The problem

Without mitigation, ~60-70% of consumer code breaks on upgrade:

```csharp
double meters = length.Meters;          // ❌ won't compile (returns QuantityValue)
double ratio = len1 / len2;             // ❌ won't compile (returns QuantityValue)
Math.Round(length.Meters, 2);           // ❌ won't compile
SomeApi(length.Value);                  // ❌ won't compile
```

---

## Option A: Add implicit conversion `QuantityValue → double` (recommended)

Change one keyword in `QuantityValue.ConvertToNumber.cs`:

```csharp
// Before:
public static explicit operator double(QuantityValue value) => value.ToDouble();
// After:
public static implicit operator double(QuantityValue value) => value.ToDouble();
```

### Why it works

The PR is clean — **no generated operator takes `double`**. Every operator uses `QuantityValue` exclusively:

```csharp
Length operator *(QuantityValue left, Length right)
Length operator *(Length left, QuantityValue right)
Length operator /(Length left, QuantityValue right)
QuantityValue operator /(Length left, Length right)
```

With bidirectional implicit conversions (`double ↔ QuantityValue`), C# overload resolution prefers user-defined operators over built-in ones when both require one implicit conversion. So `QV + double` resolves to user-defined `QV + QV`, not built-in `double + double`.

### Verified empirically (branch `agl/pr-1544-review-option-a-implicit-cast`)

https://github.com/angularsen/UnitsNet/commit/9d734dee6f9c34ea5f0012357853b6b3b3a1d792

We implemented this and built/tested the full solution:

- **Library code**: 1 line changed. **Zero compilation errors.**
- **Test code**: 101 `Assert.Equal(double, QuantityValue)` calls needed `(double)` casts to resolve xUnit overload ambiguity — a mechanical fix, not a behavioral change.
- **Build**: 0 errors, 0 warnings.
- **Tests**: **0 new failures** across ~51,900 tests.

The xUnit ambiguity is specific to `Assert.Equal`'s many overloads (including `DateTime`). Standard consumer code — method calls, assignments, operators, comparisons — is unaffected.

### Consumer code scenarios

| Scenario | Works? | Explanation |
|----------|--------|-------------|
| `double meters = length.Meters` | ✅ | Implicit conversion |
| `double ratio = len1 / len2` | ✅ | Operator returns `QV`, implicit converts |
| `Math.Round(length.Meters, 2)` | ✅ | Resolves to `Math.Round(double, int)` |
| `SomeDoubleApi(length.Meters)` | ✅ | Implicit conversion |
| `Length * double` | ✅ | `double` promotes to `QV`, uses `Length * QV` |
| `double x = double + QV` | ✅ | Operator returns `QV`, implicit converts for assignment |
| `var x = length.Meters` | ✅ | `x` is `QV` — implicitly converts wherever `double` is expected |

### Trade-offs

- `var x = length.Meters` infers `QuantityValue` — consumers see this in IDE tooltips. Not a functional issue (implicit conversion handles everything), but visible.
- Precision loss from `QuantityValue → double` is silent. This is the desired trade-off for most consumers. Those who want exact precision can stay on `QuantityValue` directly.
- Bidirectional implicit conversion is unusual in C#, but our empirical testing confirms it works cleanly here.

---

## Option B: Keep `double` public, use `QuantityValue` internally (not recommended)

Revert the public API to `double` while keeping `QuantityValue` for the internal conversion pipeline. We fully implemented this to evaluate the trade-offs.

### Verified empirically (branch `agl/pr-1544-review-option-b-double-public`)

https://github.com/angularsen/UnitsNet/commit/37d1641698ab2e2bc876ed05ad0ff9dbfabc1c40

- **608 files changed**: ~10 manual source files, CodeGen templates (~15 locations), `UnitRelations.json`, all 131 generated quantities regenerated, serialization adapters, test and benchmark code.
- **Build**: 0 library errors, 0 warnings.
- **Tests**: **2,509 failures out of 51,899 (5%).**

### The failures are not test bugs

They reveal a fundamental precision loss from `double` storage:

| Failure category | Count | Root cause |
|-----------------|-------|------------|
| `ToUnit_FromNonBaseUnit` | 1,215 | Roundtrip conversion loses precision — `double` introduces rounding at each step |
| `ToUnit_FromIQuantity` | 491 | Same root cause |
| `ToString` | 256 | Precision-sensitive formatting |
| `ConversionRoundTrip` | 68 | Direct roundtrip tests |

With `QuantityValue`, converting `3.0 feet → meters → feet` is lossless (exact rational arithmetic). With `double` storage, each conversion step introduces floating-point rounding that compounds. **This defeats a core value proposition of the PR.**

---

## Comparison

| | Option A | Option B |
|---|---|---|
| Library code changes | 1 line | 608 files |
| Build errors | 0 | 0 |
| New test failures | 0 | 2,509 (5%) |
| Precision preserved | Full | Degraded (roundtrip loss) |
| Consumer breaking changes eliminated | ~60-70% | ~60-70% |
| Future flexibility | Consumers can use `QuantityValue` directly | Requires later API change |

## Recommendation

**Option A.** One line of library code, zero new test failures, full precision preserved. It eliminates the largest consumer-facing breaking change while keeping 100% of the PR's value.

---

## Other items to address before merge

### Must do

1. **Revert `EmitDefaultValue = false` on DataMember** — `Length.Zero` serialized via DataContract should always include both Value and Unit. @angularsen flagged this.

### Should do

2. **Document DataContract serialization as a breaking change** in v6 release notes. The `_value` field type change means existing DataContract-serialized data won't deserialize. XML surrogate works; JSON surrogate blocked by [.NET runtime bug](https://github.com/dotnet/runtime/issues/100553). Recommend migrating to JsonNet or System.Text.Json packages.

### Post-merge

3. **Migration guide** for this PR's breaking changes (to be expanded for full v6 later).
4. **Test coverage for `InterfaceQuantityWithUnitTypeConverter`** — appears unused/untested.
5. **Align `QuantityValueFormatOptions` enum names** — `DecimalPrecision` vs `ExactNumber` should use consistent naming.
6. **`Configure` extension on `UnitDefinition[]`** — consider wrapper type or static method for cleaner API.

### Remaining breaking changes (regardless of QuantityValue approach)

| Change | v5 | v6 |
|--------|-----|-----|
| `As()`, `ToUnit()` | Interface methods | Extension methods (some `[Obsolete]`) |
| `UnitConverter()` constructor | Public, parameterless | Removed |
| `SetConversionFunction` / `GetConversionFunction` | Available | Removed |
| `UnitsNetSetup` constructor | Public | Private; use builder |
| DataContract serialization | `double` field | `QuantityValue` struct |
| `AbbreviatedUnitsConverter` (JsonNet) | `IReadOnlyDictionary` ctor | `UnitParser` + options |
| Default JSON precision | ~17 digits | Up to 29 digits |
| `MissingMemberHandling.Error` | Silently skipped unknowns | Correctly throws (bug fix) |
| Null `IQuantity` deserialization | Returns `.Zero` | Returns `null` |

lipchev committed:
merged from upstream and bumped the version to 6.0.0-pre102

lipchev committed:
bump the rest of the packages to 6.0.0-pre102

lipchev committed:
making the QuantityValue -> double operator implicit

lipchev committed:
removing the EmitDefaultValue = false from the quantity unit and value

lipchev committed:
bumping the samples package reference

angularsen committed:
Merge remote-tracking branch 'origin/master' into pr-1544

angularsen committed:
Merge remote-tracking branch 'origin/master' into pr-1544

# Conflicts:
#	CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs
#	Common/UnitRelations.json
#	UnitsNet.NumberExtensions.CS14/GeneratedCode/NumberToTorqueExtensions.g.cs
#	UnitsNet.NumberExtensions/GeneratedCode/NumberToTorqueExtensions.g.cs
#	UnitsNet.Tests/CustomCode/IQuantityTests.cs
#	UnitsNet.Tests/CustomQuantities/HowMuch.cs
#	UnitsNet.Tests/GeneratedCode/TestsBase/TorqueTestsBase.g.cs
#	UnitsNet/Extensions/QuantityExtensions.cs
#	UnitsNet/GeneratedCode/Quantities/AbsorbedDoseOfIonizingRadiation.g.cs
#	UnitsNet/GeneratedCode/Quantities/Acceleration.g.cs
#	UnitsNet/GeneratedCode/Quantities/AmountOfSubstance.g.cs
#	UnitsNet/GeneratedCode/Quantities/AmplitudeRatio.g.cs
#	UnitsNet/GeneratedCode/Quantities/Angle.g.cs
#	UnitsNet/GeneratedCode/Quantities/Area.g.cs
#	UnitsNet/GeneratedCode/Quantities/AreaDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/AreaMomentOfInertia.g.cs
#	UnitsNet/GeneratedCode/Quantities/AreaPerLength.g.cs
#	UnitsNet/GeneratedCode/Quantities/BitRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/BrakeSpecificFuelConsumption.g.cs
#	UnitsNet/GeneratedCode/Quantities/CoefficientOfThermalExpansion.g.cs
#	UnitsNet/GeneratedCode/Quantities/Compressibility.g.cs
#	UnitsNet/GeneratedCode/Quantities/Density.g.cs
#	UnitsNet/GeneratedCode/Quantities/DoseAreaProduct.g.cs
#	UnitsNet/GeneratedCode/Quantities/Duration.g.cs
#	UnitsNet/GeneratedCode/Quantities/DynamicViscosity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricAdmittance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricApparentEnergy.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricApparentPower.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricCapacitance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricCharge.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricChargeDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricConductance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricConductivity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricCurrent.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricCurrentDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricCurrentGradient.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricField.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricImpedance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricInductance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricPotential.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricPotentialChangeRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricReactance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricReactiveEnergy.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricReactivePower.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricResistance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricResistivity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricSurfaceChargeDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ElectricSusceptance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Energy.g.cs
#	UnitsNet/GeneratedCode/Quantities/EnergyDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/Entropy.g.cs
#	UnitsNet/GeneratedCode/Quantities/FluidResistance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Force.g.cs
#	UnitsNet/GeneratedCode/Quantities/ForceChangeRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/ForcePerLength.g.cs
#	UnitsNet/GeneratedCode/Quantities/Frequency.g.cs
#	UnitsNet/GeneratedCode/Quantities/FuelEfficiency.g.cs
#	UnitsNet/GeneratedCode/Quantities/HeatFlux.g.cs
#	UnitsNet/GeneratedCode/Quantities/HeatTransferCoefficient.g.cs
#	UnitsNet/GeneratedCode/Quantities/Illuminance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Impulse.g.cs
#	UnitsNet/GeneratedCode/Quantities/Information.g.cs
#	UnitsNet/GeneratedCode/Quantities/Irradiance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Irradiation.g.cs
#	UnitsNet/GeneratedCode/Quantities/Jerk.g.cs
#	UnitsNet/GeneratedCode/Quantities/KinematicViscosity.g.cs
#	UnitsNet/GeneratedCode/Quantities/LeakRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/Length.g.cs
#	UnitsNet/GeneratedCode/Quantities/Level.g.cs
#	UnitsNet/GeneratedCode/Quantities/LinearDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/LinearPowerDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/Luminance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Luminosity.g.cs
#	UnitsNet/GeneratedCode/Quantities/LuminousFlux.g.cs
#	UnitsNet/GeneratedCode/Quantities/LuminousIntensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/MagneticField.g.cs
#	UnitsNet/GeneratedCode/Quantities/MagneticFlux.g.cs
#	UnitsNet/GeneratedCode/Quantities/Magnetization.g.cs
#	UnitsNet/GeneratedCode/Quantities/Mass.g.cs
#	UnitsNet/GeneratedCode/Quantities/MassConcentration.g.cs
#	UnitsNet/GeneratedCode/Quantities/MassFlow.g.cs
#	UnitsNet/GeneratedCode/Quantities/MassFlux.g.cs
#	UnitsNet/GeneratedCode/Quantities/MassFraction.g.cs
#	UnitsNet/GeneratedCode/Quantities/MassMomentOfInertia.g.cs
#	UnitsNet/GeneratedCode/Quantities/Molality.g.cs
#	UnitsNet/GeneratedCode/Quantities/MolarEnergy.g.cs
#	UnitsNet/GeneratedCode/Quantities/MolarEntropy.g.cs
#	UnitsNet/GeneratedCode/Quantities/MolarFlow.g.cs
#	UnitsNet/GeneratedCode/Quantities/MolarMass.g.cs
#	UnitsNet/GeneratedCode/Quantities/Molarity.g.cs
#	UnitsNet/GeneratedCode/Quantities/Permeability.g.cs
#	UnitsNet/GeneratedCode/Quantities/Permittivity.g.cs
#	UnitsNet/GeneratedCode/Quantities/PorousMediumPermeability.g.cs
#	UnitsNet/GeneratedCode/Quantities/Power.g.cs
#	UnitsNet/GeneratedCode/Quantities/PowerDensity.g.cs
#	UnitsNet/GeneratedCode/Quantities/PowerRatio.g.cs
#	UnitsNet/GeneratedCode/Quantities/Pressure.g.cs
#	UnitsNet/GeneratedCode/Quantities/PressureChangeRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/RadiationEquivalentDose.g.cs
#	UnitsNet/GeneratedCode/Quantities/RadiationEquivalentDoseRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/RadiationExposure.g.cs
#	UnitsNet/GeneratedCode/Quantities/Radioactivity.g.cs
#	UnitsNet/GeneratedCode/Quantities/Ratio.g.cs
#	UnitsNet/GeneratedCode/Quantities/RatioChangeRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/ReciprocalArea.g.cs
#	UnitsNet/GeneratedCode/Quantities/ReciprocalLength.g.cs
#	UnitsNet/GeneratedCode/Quantities/RelativeHumidity.g.cs
#	UnitsNet/GeneratedCode/Quantities/RotationalAcceleration.g.cs
#	UnitsNet/GeneratedCode/Quantities/RotationalSpeed.g.cs
#	UnitsNet/GeneratedCode/Quantities/RotationalStiffness.g.cs
#	UnitsNet/GeneratedCode/Quantities/RotationalStiffnessPerLength.g.cs
#	UnitsNet/GeneratedCode/Quantities/Scalar.g.cs
#	UnitsNet/GeneratedCode/Quantities/SolidAngle.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificEnergy.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificEntropy.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificFuelConsumption.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificVolume.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificWeight.g.cs
#	UnitsNet/GeneratedCode/Quantities/Speed.g.cs
#	UnitsNet/GeneratedCode/Quantities/StandardVolumeFlow.g.cs
#	UnitsNet/GeneratedCode/Quantities/Temperature.g.cs
#	UnitsNet/GeneratedCode/Quantities/TemperatureChangeRate.g.cs
#	UnitsNet/GeneratedCode/Quantities/TemperatureDelta.g.cs
#	UnitsNet/GeneratedCode/Quantities/TemperatureGradient.g.cs
#	UnitsNet/GeneratedCode/Quantities/ThermalConductivity.g.cs
#	UnitsNet/GeneratedCode/Quantities/ThermalInsulance.g.cs
#	UnitsNet/GeneratedCode/Quantities/ThermalResistance.g.cs
#	UnitsNet/GeneratedCode/Quantities/Torque.g.cs
#	UnitsNet/GeneratedCode/Quantities/Turbidity.g.cs
#	UnitsNet/GeneratedCode/Quantities/VitaminA.g.cs
#	UnitsNet/GeneratedCode/Quantities/Volume.g.cs
#	UnitsNet/GeneratedCode/Quantities/VolumeConcentration.g.cs
#	UnitsNet/GeneratedCode/Quantities/VolumeFlow.g.cs
#	UnitsNet/GeneratedCode/Quantities/VolumeFlowPerArea.g.cs
#	UnitsNet/GeneratedCode/Quantities/VolumePerLength.g.cs
#	UnitsNet/GeneratedCode/Quantities/VolumetricHeatCapacity.g.cs
#	UnitsNet/GeneratedCode/Quantities/WarpingMomentOfInertia.g.cs
#	UnitsNet/IArithmeticQuantity.cs
#	UnitsNet/IQuantity.cs

angularsen committed:
Fix conversion tests for equivalent target units

angularsen committed:
Merge remote-tracking branch 'origin/master' into pr-1544

# Conflicts:
#	UnitsNet.NumberExtensions.CS14/GeneratedCode/NumberToSpecificVolumeExtensions.g.cs
#	UnitsNet.NumberExtensions.CS14/GeneratedCode/NumberToVolumeExtensions.g.cs
#	UnitsNet.NumberExtensions/GeneratedCode/NumberToSpecificVolumeExtensions.g.cs
#	UnitsNet.NumberExtensions/GeneratedCode/NumberToVolumeExtensions.g.cs
#	UnitsNet.Tests/GeneratedCode/TestsBase/SpecificVolumeTestsBase.g.cs
#	UnitsNet.Tests/GeneratedCode/TestsBase/VolumeTestsBase.g.cs
#	UnitsNet/GeneratedCode/Quantities/SpecificVolume.g.cs
#	UnitsNet/GeneratedCode/Quantities/Volume.g.cs

angularsen committed:
Address remaining fractional quantity review blockers

angularsen commented on UnitsNet.Serialization.JsonNet/QuantityValueFormatOptions.cs:
```diff
@@ -0,0 +1,101 @@
+    ///     This number is represented with up to 17 significant digits. Numbers such as "1/3" would be represented as
+    ///     "0.3333333333333333".
+    /// </remarks>
+    DoublePrecision,
```
Final implementation: `DecimalPrecision` remains the default serialization format and writes up to 29 significant digits (`G29`). `DoublePrecision` remains an explicit 17-significant-digit (`G17`) compatibility format. A regression test now verifies that the 17th digit is preserved instead of accidentally using the default `G16` formatter.

angularsen replied:
Final result: the serialization option remains `DecimalPrecision`, parallel to `DoublePrecision`. `DecimalPrecision` writes `G29`; `DoublePrecision` now explicitly writes `G17`, with a regression test covering the 17th significant digit. Both remain available for their distinct precision and compatibility requirements.

angularsen replied:
The `DoublePrecision` compatibility format now explicitly writes `QuantityValue` with `G17`, matching its documented limit of 17 significant digits instead of accidentally using the default `G16` formatter. A regression test covers a value whose 17th digit would previously have changed.

angularsen commented on UnitsNet.Serialization.JsonNet/QuantityValueFormatOptions.cs:
```diff
@@ -0,0 +1,101 @@
+    ///     This format is ideal for scenarios where the exact representation of a value is required for round-trip
+    ///     serialization and deserialization.
+    /// </remarks>
+    RoundTripping,
```
Final naming: serialization uses `DecimalPrecision`, deserialization uses `ExactNumber`, and the exact round-trip option is `RoundTrip` in both enums. The defaults are `DecimalPrecision` for writing and `ExactNumber` for reading. The direction-specific names are intentional: the writer emits a bounded decimal representation, while the reader parses every digit of a JSON number into the numeric `QuantityValue` without retaining the token's original spelling.

angularsen replied:
Final naming: serialization uses `DecimalPrecision`, deserialization uses the more accurate `ExactNumber`, and the exact round-trip option is `RoundTrip` in both option enums.

angularsen replied:
Final decision: the exact round-trip option is named `RoundTrip` in both Json.NET option enums, with the helpers, tests, and documentation aligned.

angularsen commented on UnitsNet.Serialization.JsonNet/QuantityValueFormatOptions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <summary>
+    ///     Deserializes the quantity value as exact number in decimal notation.
+    /// </summary>
+    ExactNumber,
```
Final implementation intentionally uses direction-specific names: serialization uses `DecimalPrecision` (`G29`), while deserialization uses `ExactNumber` because it parses every digit of the JSON number without conversion through `double` or `decimal`. The double-related names also describe their behavior: `DoublePrecision` writes `G17`, while `RoundedDouble` restores the previous 15-significant-digit reading behavior.

angularsen replied:
Final correction after checking the implementation: deserialization is named `ExactNumber`, not `DecimalPrecision`. It reads the JSON number as text to avoid `double`/`decimal` coercion and parses every digit into the numeric `QuantityValue`; it does not retain the token's original spelling. `RoundedDouble` remains distinct because it deliberately restores the previous 15-significant-digit behavior.

angularsen replied:
Final correction after checking the implementation rather than forcing symmetric names: serialization remains `DecimalPrecision`, while deserialization is restored to `ExactNumber`. `ExactNumber` reads the JSON number as text to avoid `double`/`decimal` coercion and parses every digit into the numeric `QuantityValue`; it does not retain the token's original spelling. The differing names now reflect genuinely different direction-specific behavior.

angularsen commented on UnitsNet.Serialization.JsonNet.Tests/Value/QuantityValueFractionalNotationConverterTests.cs:
```diff
@@ -0,0 +1,299 @@
+    public void Serialize_ReturnsTheExpectedResult()
+    {
+        // Arrange
+        var json = """{"N":42,"D":10}""";
```
The serializers now offer both representations. `QuantityValueFractionalNotationConverter` retains the explicit numerator/denominator object, while `QuantityValueMixedNotationConverter` writes finite values as readable decimal numbers and non-terminating values as strings such as `"1/3"`, preserving exact round trips. The serialization guide now calls out the mixed converter for this use case.

angularsen commented on UnitsNet.Serialization.JsonNet.Tests/AbbreviatedUnitsConverterWithCustomFormatOptionsTests.cs:
```diff
@@ -0,0 +1,17 @@
+                new QuantityValueFormatOptions(QuantityValueSerializationFormat.Custom, QuantityValueDeserializationFormat.Custom)),
+            new QuantityValueFractionalNotationConverter())
+    {
+    }
```
This class configures the custom `QuantityValueFractionalNotationConverter` and inherits the shared serialization test suite from `JsonNetSerializationTestsBase` / `SerializationTestsBase`; it is therefore exercised even though this derived class contains no test methods of its own. The current Json.NET run passes 213 tests on each target framework.

angularsen commented on UnitsNet.Serialization.SystemTextJson/InterfaceQuantityWithUnitTypeConverter.cs:
```diff
@@ -0,0 +1,303 @@
+[RequiresUnreferencedCode(
+    "If some of the generic arguments are annotated (either with DynamicallyAccessedMembersAttribute, or generic constraints), trimming can't validate that the requirements of those annotations are met.")]
+#endif
+public class InterfaceQuantityWithUnitTypeConverter : JsonConverter<IQuantity>
```
Confirmed after tracing its history and usage: `InterfaceQuantityWithUnitTypeConverter` was introduced in the initial bulk commit, had no consumers, tests, or documentation, and duplicated the supported converter composition. It has been removed in the pending cleanup rather than turning an experimental API into supported surface area.

angularsen commented on UnitsNet.Tests/Serialization/Json/DefaultDataContractJsonSerializerTests.cs:
```diff
@@ -59,11 +56,11 @@ public void DoubleQuantity_SerializedWithDoubleValueAndunitInt()
             var expectedUnit = MassUnit.Milligram;
             var unitInt = (int)expectedUnit;
-            var json = $"{{\"Value\":1.2,\"Unit\":{unitInt}}}";
+            var json = $$$"""{"Value":{"N":{"_bits":null,"_sign":12},"D":{"_bits":null,"_sign":10}},"Unit":{{{unitInt}}}}""";
```
The migration decision is now explicit in the v6 upgrade and serialization guides. `DataContractJsonSerializer` cannot apply the surrogate provider to nested `QuantityValue` instances on .NET 10 and earlier due to [dotnet/runtime#100553](https://github.com/dotnet/runtime/issues/100553). This was fixed in .NET 11 Preview 6 by [dotnet/runtime#118257](https://github.com/dotnet/runtime/pull/118257), so applications running on .NET 11 Preview 6 or later can configure `QuantityValueSurrogateSerializationProvider`. The serialized representation is still a v6 schema change, and Json.NET or System.Text.Json remains the recommended cross-runtime option.

angularsen commented on UnitsNet.Tests/Serialization/Xml/DataContractSerializerTests.cs:
```diff
@@ -37,10 +36,47 @@ protected override T DeserializeObject<T>(string xml)
-            var expectedXml = $"<Mass {Namespace} {XmlSchema}><Value>1.2</Value><Unit>Milligram</Unit></Mass>";
+            var numeratorFormat = $"<N xmlns:a={NumericsNamespace}><a:_bits i:nil=\"true\" xmlns:b={ArraysNamespace}/><a:_sign>12</a:_sign></N>";
+            var denominatorFormat = $"<D xmlns:a={NumericsNamespace}><a:_bits i:nil=\"true\" xmlns:b={ArraysNamespace}/><a:_sign>10</a:_sign></D>";
+            var expectedXml = $"<Mass {Namespace} {XmlSchema}><Value>{numeratorFormat}{denominatorFormat}</Value><Unit>Milligram</Unit></Mass>";
```
Addressed in the migration documentation. The guide now explains that the default XML representation exposes the internal `BigInteger` fields, shows how to configure `QuantityValueSurrogateSerializationProvider`, and documents its compact, stable numerator/denominator representation. This is an explicit v6 schema migration rather than silent compatibility with the old scalar value.

angularsen commented on UnitsNet.Tests/CustomCode/QuantityInfos/Builder/QuantityInfoBuilderExtensionsTest.cs:
```diff
@@ -0,0 +1,91 @@
+            .ToList();
+        var customMeterDefinition = new UnitDefinition<LengthUnit>(LengthUnit.Meter, "UpdatedMeter", "UpdatedMeters", BaseUnits.Undefined);
+
+        var result = unitDefinitions.Configure(LengthUnit.Meter, _ => customMeterDefinition).ToList();
```
Reviewed again and decided to keep the `Configure` extension API as-is. It provides a concise transformation over the unit-definition collection, and introducing a wrapper or static helper would add API surface without a clear behavioral benefit.

angularsen committed:
Merge remote-tracking branch 'origin/master' into pr-1544
